### PR TITLE
For 1811851: Update message notification worker to handle when a user clicks or dismisses a Notification.

### PR DESCRIPTION
**Pull request by t-p-white**
Fixes [EXP-3068](https://mozilla-hub.atlassian.net/browse/EXP-3068).

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation

Used by GitHub Actions.

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/gleanplumb/NimbusMessagingController.kt [hunk 1]
```diff
@@ -68,17 +102,6 @@ class NimbusMessagingController(
         return action.toDeepLinkSchemeUri()
     }
 
-    /**
```
Comment (t-p-white) [inline app/src/main/java/org/mozilla/fenix/gleanplumb/NimbusMessagingController.kt:71]: moved to line 64

File: app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt [hunk 2]
```diff
@@ -57,46 +60,81 @@ class MessageNotificationWorker(
...
-                message.data.text,
-                pendingIntent,
-            )
-        }
+    private fun createOnClickPendingIntent(
+        context: Context,
+        nimbusMessagingController: NimbusMessagingController,
+        message: Message,
+    ): PendingIntent {
+        val intent = nimbusMessagingController.processMessageActionToIntent(message)
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:116]: IIUC, the action's UUID is being generated and recorded with Glean as if the message was clicked on.

File: app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt [hunk 3]
```diff
@@ -123,3 +161,32 @@ class MessageNotificationWorker(
...
+
+            val messageMetadata =
+                JSONObject(intent?.getStringExtra(MESSAGE_METADATA)!!).toMetadata()
+
+            nimbusMessagingController.onMessageDismissed(messageMetadata)
+        }
+
+        return START_REDELIVER_INTENT
+    }
+}
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:192]: I love this use of a service. 

I think I'd like to see this (or similar service) catch the `onMessageClicked`, process the action to make a new Intent, and then forward on.

File: app/src/main/java/org/mozilla/fenix/home/intent/HomeDeepLinkIntentProcessor.kt [hunk 4]
```diff
@@ -19,26 +21,52 @@ import org.mozilla.fenix.GlobalDirections
...
+    private val nimbusMessagingController: NimbusMessagingController = NimbusMessagingController(
+        activity.components.analytics.messagingStorage,
+    ),
 ) : HomeIntentProcessor {
     private val logger = Logger("DeepLinkIntentProcessor")
 
     override fun process(intent: Intent, navController: NavController, out: Intent): Boolean {
-        val scheme = intent.scheme?.equals(BuildConfig.DEEP_LINK_SCHEME, ignoreCase = true) ?: return false
+        val scheme =
+            intent.scheme?.equals(BuildConfig.DEEP_LINK_SCHEME, ignoreCase = true) ?: return false
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/home/intent/HomeDeepLinkIntentProcessor.kt:44]: Note that we may have messages with URIs that take the user to other apps, e.g. 

```
market://launch?id=<package_name>
```

In such cases, we either can't deal with it here, or we don't get to here.

**Review by jhugman — COMMENTED** [review]
I think this is beginning to take shape!

I would strongly urge you to avoid touching the `HomeDeepLinkIntentProcessor`, and add logic to the `NotificationDismissedService` (perhaps renaming to something like `NotificationInteractorService`) to unpack the action/uuid/Glean before forwarding on.

File: app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt (hunk 2 above)
Comment (t-p-white) [reply app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:116]: good spot! 🙌

File: app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt (hunk 3 above)
Comment (t-p-white) [reply app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:192]: Nice suggestion 👍

File: app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt [hunk 5]
```diff
@@ -37,66 +45,113 @@ class MessageNotificationWorker(
...
+        nimbusMessagingController: NimbusMessagingController,
+        messageId: String,
+    ): PendingIntent? = withContext(coroutineScope.coroutineContext) {
+        val message =
+            nimbusMessagingController.getMessage(messageId) ?: return@withContext null
+
+        // Update message as 'clicked'.
+        nimbusMessagingController.onMessageClicked(message.metadata)
+
+        val intent = nimbusMessagingController.getIntentForMessageAction(message.action)
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:117]: Both of these calls need to be done _after_ the user has clicked on the message.

I wonder if we can use the `IntentReceiverActivity` with a notification specific `IntentProcessor`?

I think we need advice from @Amejia481!

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:117]: I guess it depends how we want to handle a notification click, but all the actions that we are getting are from nimbus are being handled on [HomeDeepLinkIntentProcessor](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/home/intent/HomeDeepLinkIntentProcessor.kt)  :) 

See `toDeepLinkSchemeUri`
https://github.com/mozilla-mobile/fenix/blob/249ddef570c14ea748524fadceb9d55446e28410/app/src/main/java/org/mozilla/fenix/gleanplumb/NimbusMessagingController.kt#L101-L111

**Review by jhugman — APPROVED** [review]
I am happy to approve this. From the Nimbus p.o.v. it appears nice and clean. Good job.

✅ ✅

**Pull request merged**